### ISSUE TITLE
Capturing things routed to BlackHoleCluster or 502'd when REGISTRY_ONLY

### ISSUE THREAD

**Issue by Stono** [istio/istio#7669]
Hey all, 
We have istio-proxy intercept all outbound traffic, at the moment of a pod was to connect to say, `cnn.com`, and we didn't have a ServiceEntry defined - then istio-proxy would (rightly) 404.

I would *love it* if requests to services that were not defined as service entries such as this were logged in some way as prometheus metrics, just like requests to correctly defined ServiceEntries are.

This would enable us to quickly identify:

  - Applications attempting to make connections that are potentially malicious / unexpected
  - Enable us to debug legacy apps we've moved over that frankly, we're not 100% sure what they're talking out to

Thanks
Karl

**Comment by mandarjog** [comment]
This is a great suggestion.
We can include `NR` (and other envoy statuses) in the report call. We will expose these in metrics.

**Comment by Stono** [comment]
Brill @mandarjog - I look forward to it!

**Comment by Stono** [comment]
not stale....

**Comment by akpsgit** [comment]
Great suggestion. Will be very useful.
By the way, can't this be done using outboundTrafficPolicy? from description:
    // Set the default behavior of the sidecar for handling outbound traffic from the application.
    // While the default mode should work out of the box, if your application uses one or more external services that
    // are not known apriori, setting the policy to ALLOW_ANY will cause the sidecars to route traffic to the any
    // requested destination.
    // Users are strongly encouraged to use ServiceEntries to explicitly declare any external dependencies,
    // instead of using allow_any.

**Comment by howardjohn** [comment]
Does https://github.com/istio/istio/pull/12916 fix this? No prometheus metrics but you get a 502 error rather than a 404

**Comment by Stono** [comment]
No not really, this is specifically around visibility from a signals perspective of the issue.

**Comment by nrjpoddar** [comment]
This is something that we also need both for our usage and our customers. Is there an agreement on how to add this functionality? I can add the code needed if we agree on the approach.

**Comment by nrjpoddar** [comment]
@Stono I think the behavior is a little different if you do HTTPS for external services and  if you don't have a ServiceEntry defined and you're by default blocking all external access, you don't get a 404 NR.

**Comment by Stono** [comment]
Yeah this thread is about capturing 502'd stuff in registry only.  

Fundamentally I just want more metrics capturing failed requests ha

**Comment by nrjpoddar** [comment]
Proposal from discussion on slack:

Update pilot to add the mixer filter for the `0.0.0.0:15001` listener before it gets routed to the BlackHole cluster. I'm hoping by doing that the mixer filter can add some context and we can generate metrics based on those attributes. Here's the Envoy config where this change will be made:

```json

          "version_info": "2019-04-16T18:23:50Z/12",
          "listener": {
            "name": "virtual",
            "address": {
              "socket_address": {
                "address": "0.0.0.0",
                "port_value": 15001
              }
            },
            "filter_chains": [
              {
                "filters": [
                  {
                    "name": "envoy.tcp_proxy",
                    "config": {
                      "cluster": "BlackHoleCluster",
                      "stat_prefix": "BlackHoleCluster"
                    }
                  }
                ]
              }
            ],
            "use_original_dst": true
          },
          "last_updated": "2019-04-16T18:23:55.335Z"
        }
```

@mandarjog @douglas-reid will this approach work? open to suggestions.

**Comment by douglas-reid** [comment]
@nrjpoddar do you have an easy way to test with some hand-altered config?  It seems, at first glance, to be reasonable -- we'd just have to agree on what values to place in the attributes, etc.

**Comment by michaelkipper** [comment]
As discussed on Slack, this is just as much applicable to the Passthrough cluster as it is to the Blackhole cluster.

**Comment by nrjpoddar** [comment]
@douglas-reid @howardjohn @kyessenov  I started working on this but currently kind of stuck. I can add the logic to add a mixer filter in the filter chain for this listener but as it's a TCP filter the attributes that we can export is not rich.

In particular, you cannot get the DNS name of the service that the application is trying to reach as that information is lost by the time it gets to the mixer TCP filter. At best, we can export `destination.ip` but really what we need in this case is a way to surface the DNS name of the service. For non BlackHole cluster cases, the destination service is provided via configuration from Pilot to Envoy https://github.com/istio/istio/blob/master/pilot/pkg/networking/plugin/mixer/mixer.go#L531 which you cannot do for this use case as there's no fixed destination service known apriori.

I would appreciate any help/direction if anyone knows a better way to solve this.

**Comment by nrjpoddar** [comment]
I have thought about various ways to get around this problem of getting the DNS name as part of telemetry. Here are the options:

* Envoy intercepts DNS packets and keeps a mapping of resolved IPs to names. Pretty big change and has impact outside of this issue.
* Envoy mixer filter performs reverse DNS lookup and reports the name as part of telemetry. It's possible that the reverse DNS lookup gives a different answer as compared to what the application used. Additionally, there are performance issues that might crop up as a result of reverse lookup.
* Do protocol sniffing in Envoy (which is already proposed in Networking group) for BlackHole cluster. In this case if you understand the protocol (HTTP 1.1/2.0, gRPC) use the Host header, else only report the DestinationIP as part of TCP attributes.

None of these ideas are without broader implications and far from ideal.

**Comment by Stono** [comment]
Yeah the lesser of all the evils there is likely the third bullet point.  But they're all a "gold standard" solution.

There is an interim I feel, where you simply track the ip, operators could quite easily correlate this to running applications on their cluster should they be investigating a problem.

**Comment by douglas-reid** [comment]
@nrjpoddar i'm hoping others chime in here, but I think just having _any_ information is an improvement. So while the TCP attribute set may be more limited, it at least gives us _some_ insight into behavior. Once that is in place, we can work to improve (most likely through the protocol sniffing, etc.).

**Comment by nrjpoddar** [comment]
Thanks @stono & @douglas-reid for your inputs. Yes we can start with just IP level information but I wanted to make sure everyone understood that as the current architecture stands it would be difficult to get the DNS name out. I will wait and see if anyone else has a better/low impact solution, if not we will start with just the Destination IP and other TCP level attributes.

**Comment by douglas-reid** [comment]
@nrjpoddar one additional thing we could do is to introduce an additional attribute that indicates "blackhole" or "passthru" cluster was hit. and use that to either define new metrics or augment the existing TCP metrics. that would enable easier alerting / monitoring of unexpected traffic.

**Comment by kyessenov** [comment]
We could leave destination service un-assigned, which would imply a blackhole destination. We do need to set source namespace for the config to resolve properly.

**Comment by douglas-reid** [comment]
For TCP, we should have `source.namespace`. I'd lean towards adding an explicit value vs. relying on the absence of a value for indication.

**Comment by nrjpoddar** [comment]
Yes, source level information should be available.

I was thinking about adding BlackHole or Passthrough and I don't think overloading destination service is the correct thing to do here. We can add an extra label (i.e. attribute) which would indicate Envoy/Istio cluster name which can be same or different from the destination service name. 

Open to suggestions on name of this new label and concerns if we add it i.e. dimensionality explosion and usefulness of this label in scenarios other than BlackHole/Passthrough.

**Comment by nrjpoddar** [comment]
Filed https://github.com/envoyproxy/envoy/issues/7725 for figuring out a way to do this correctly.

**Comment by nrjpoddar** [comment]
Fixed via https://github.com/istio/istio/pull/16354

Latest istio master build should now have telemetry for BlackHole cluster. @Stono try it out and re-open if you use case is still not met.

**Comment by Stono** [comment]
I don't easily have the facility to test master @nrjpoddar unfortunately, will this be released in a 1.2.x release or is it 1.3.x?

**Comment by nrjpoddar** [comment]
It will be part of release 1.3 as this https://github.com/istio/istio/pull/16422 got merged yesterday. When the first RC comes out (which should be soon) and if you're going to qualify it please verify this functionality too. Thanks!
